feat(data): appsync apollo extensions (#7930)
aws-amplify/docs: src/pages/[platform]/build-a-backend/data/aws-appsync-apollo-extensions/index.mdx

  title: 'AWS AppSync Apollo Extensions',
  description:
    'AWS AppSync Apollo Extensions',
  platforms: [
    'android',
    'swift',
  ]
};

export function getStaticPaths() {
  return getCustomStaticPath(meta.platforms);
}

export function getStaticProps(context) {
  return {
    props: {
      platform: context.params.platform,
      meta
    }
  };
}

AWS AppSync Apollo Extensions provides a seamless way to connect to your AWS AppSync while using the Apollo client. Apollo client is an open-source GraphQL client. 

<InlineFilter filters={["swift"]}>

To learn more about Apollo, see https://www.apollographql.com/docs/ios/.

</InlineFilter>

<InlineFilter filters={["android"]}>

To learn more about Apollo, see https://www.apollographql.com/docs/kotlin.

</InlineFilter>

### Features

AWS AppSync Apollo Extensions provide AWS AppSync authorizers to be used with the Apollo client to make it simple to configure runtime interceptors and apply the correct authorization payloads to your GraphQL operations.

The Amplify library provides components to facilitate configuring the authorizers with Apollo client by providing configuration values to connect to your Amplify Data backend.

> <InlineFilter filters={["swift"]}>

Reviewer: Need Android version of this section.

Author: Added android section, can you confirm the name? I have set it roughly as `com.amplifyframework:aws-appsync-apollo-extensions-android:1.0.0`

Author: Updated

```

</InlineFilter>

You can provide your own custom `fetchLatestAuthToken` provider for **AWS_LAMBDA** and **OPENID_CONNECT** auth modes. 

#### AWS_IAM

<InlineFilter filters={["swift"]}>

If you are using Amplify Auth, you can use the following method for AWS_IAM auth

```swift
import AWSCognitoAuthPlugin
import AWSAppSyncApolloExtensions

let authorizer = IAMAuthorizer(
    signRequest: AWSCognitoAuthPlugin.createAppSyncSigner(
        region: "[REGION]"))
```

</InlineFilter>

<InlineFilter filters={["android"]}>

If you are using Amplify Auth, you can use the following method for AWS_IAM auth

```kotlin
val authorizer = IamAuthorizer { ApolloAmplifyConnector.signAppSyncRequest(it, "us-east-1") }
```

</InlineFilter>

### Connecting Amplify Data to Apollo client

Before you begin, you will need an Amplify Data backend deploy. To get started, see [Set up Data](/[platform]/swift/build-a-backend/data/set-up-data/). 

Once you have deployed your backend and created the `amplify_outputs.json` file, you can use Amplify library to read and retrieve your configuration values with the following steps:

> <InlineFilter filters={["swift"]}>

Reviewer: Need Android version of this section

Author: Added the Android section, which is 
1. the gradle for the second package, could you update to the final name?
2. a code example , need to add this

Author: Updated dependency name

Author: Updated code snippet

Reviewer 2: missing the

*  `owner` strategy, `userPool` authMode, **AuthTokenAuthorizer**

If you define multiple authorization strategies on a single model, you will have to create separate Apollo client instances for each Authorizer that you want to use in your app.

### Downloading the AWS AppSync schema

The schema is used by Apollo’s code generation tool to generate API code that helps you execute GraphQL operations. To retrieve your AWS AppSync schema:

<InlineFilter filters={["swift"]}>
1. Navigate to your API on the [AWS AppSync console](https://console.aws.amazon.com/appsync/home)
2. On the left side, select Schema
3. When viewing your schema, there should a “Export schema” drop down. Select this and download the `schema.json` file.
4. Add this file to your project as directed by [Apollo Code Generation documentation](https://www.apollographql.com/docs/ios/code-generation/introduction) 
</InlineFilter>

<InlineFilter filters={["android"]}>
1. Navigate to your API on the [AWS AppSync console](https://console.aws.amazon.com/appsync/home)
2. On the left side, select Schema
3. When viewing your schema, there should a “Export schema” drop down. Select this and download the `schema.json` file.

> 4. Add this file to your project as directed by [Apollo Code Generation documentation](https://www.apollographql.com/docs/ios/code-generation/introduction) 

Reviewer: ```suggestion
4. Add this file to your project as directed by [Apollo documentation](https://www.apollographql.com/docs/kotlin/advanced/plugin-recipes#specifying-the-schema-location) 
```

    let authorizer = /* your Authorizer */
    let interceptor = AppSyncInterceptor(authorizer)

    let interceptorProvider = DefaultPrependInterceptorProvider(interceptor: interceptor,
                                                                store: store)
    let transport = RequestChainNetworkTransport(interceptorProvider: interceptorProvider,
                                                    endpointURL: configuration.endpoint)

    // 2. Create the AWS AppSync compatible websocket client
    let websocket = AppSyncWebSocketClient(endpointURL: configuration.endpoint,
                                            authorizer: authorizer)
    // 3. Add it to the WebSocketTransport
    let webSocketTransport = WebSocketTransport(websocket: websocket)
    // 4. Create a SplitNetworkTransport
    let splitTransport = SplitNetworkTransport(
        uploadingNetworkTransport: transport,
        webSocketNetworkTransport: webSocketTransport
    )
    // 5. Pass the SplitNetworkTransport to the ApolloClient
    return ApolloClient(networkTransport: splitTransport, store: store)
}
```

</InlineFilter>

<InlineFilter filters={["android"]}>

```kotlin

> // Create the network transport
> val networkTransport = WebSocketNetworkTransport.Builder()
>     .protocol(AppSyncProtocol.Factory(endpoint, authorizer))
>     .build()
> 
> // Use the network transport when creating the Apollo Client
> val apolloClient = ApolloClient.Builder()
>     .subscriptionNetworkTransport(networkTransport)
>     .build()

Reviewer: ```suggestion
val endpoint = AppSyncEndpoint("<your_appsync_endpoint>")
val authorizer = /* your Authorizer */

val apolloClient = ApolloClient.Builder()
    .appSync(endpoint, authorizer)
    .build()
```

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {
  title: 'AWS AppSync Apollo Extensions',
  description:
    'AWS AppSync Apollo Extensions',

>   platforms: [
>     'android',
>     'angular',
>     'flutter',
>     'javascript',
>     'nextjs',
>     'react',
>     'react-native',
>     'swift',
>     'vue'
>   ]

Reviewer 3: ```suggestion
  platforms: [
    'android',
    'swift',
  ]
```

};

export function getStaticPaths() {
  return getCustomStaticPath(meta.platforms);
}

export function getStaticProps(context) {
  return {
    props: {
      platform: context.params.platform,
      meta
    }
  };
}

> AWS AppSync Apollo Extensions provides a seamless way to connect to your AWS AppSync while using the Apollo client. Apollo client is an open-source GraphQL client. 

Reviewer 3: can we add a link to "Apollo client"?

Reviewer 3: disregard I see the following notes, that looks good

<InlineFilter filters={["swift"]}>

To learn more about Apollo, see https://www.apollographql.com/docs/ios/.

</InlineFilter>

<InlineFilter filters={["android"]}>

To learn more about Apollo, see https://www.apollographql.com/docs/kotlin.

</InlineFilter>

### Features

AWS AppSync Apollo Extensions provide AWS AppSync authorizers to be used with the Apollo client to make it simple to configure runtime interceptors and apply the correct authorization payloads to your GraphQL operations.

The Amplify library provides components to facilitate configuring the authorizers with Apollo client by providing configuration values to connect to your Amplify Data backend.

<InlineFilter filters={["swift"]}>

### Install the AWS AppSync Apollo library

Add AWS AppSync Apollo Extensions into your project using Swift Package Manager.

Enter its GitHub URL (`https://github.com/aws-amplify/aws-appsync-apollo-extensions-swift`), select **Up to Next Major Version** and click **Add Package**

* Select the following libraries:
    * **AWSAppSyncApolloExtensions**

</InlineFilter>

### Connecting to AWS AppSync with Apollo client

> AWS AppSync supports the following authorization modes (https://docs.aws.amazon.com/appsync/latest/devguide/security-authz.html)

Reviewer 3: ```suggestion
AWS AppSync supports the following [authorization modes](https://docs.aws.amazon.com/appsync/latest/devguide/security-authz.html)
```

func getUserPoolAccessToken() async throws -> String {
    let authSession = try await Amplify.Auth.fetchAuthSession()
    if let result = (authSession as? AuthCognitoTokensProvider)?.getCognitoTokens() {
        switch result {
        case .success(let tokens):
            return tokens.accessToken
        case .failure(let error):
            throw error
        }
    }
    throw AuthError.unknown("Did not receive a valid response from fetchAuthSession for get token.")
}
```

Then create the AuthTokenAuthorizer with this method.

```swift
import AWSAppSyncApolloExtensions

let authorizer = AuthTokenAuthorizer(fetchLatestAuthToken: getUserPoolAccessToken)
let interceptor = AppSyncInterceptor(authorizer)
```
</InlineFilter>

<InlineFilter filters={["android"]}>

Create the AuthTokenAuthorizer with this method.

```kotlin
let authorizer = AuthTokenAuthorizer(fetchLatestAuthToken: getLatestTokenFromAmplify)
```

</InlineFilter>

You can provide your own custom `fetchLatestAuthToken` provider for **AWS_LAMBDA** and **OPENID_CONNECT** auth modes. 

#### AWS_IAM

<InlineFilter filters={["swift"]}>

If you are using Amplify Auth, you can use the following method for AWS_IAM auth

> ```

Reviewer 3: ````suggestion
```swift
````

import AWSCognitoAuthPlugin
import AWSAppSyncApolloExtensions

let authorizer = IAMAuthorizer(
    signRequest: AWSCognitoAuthPlugin.createAppSyncSigner(
        region: "[REGION]"))
```

</InlineFilter>

<InlineFilter filters={["android"]}>

If you are using Amplify Auth, you can use the following method for AWS_IAM auth

> ```

Reviewer 3: ````suggestion
```kotlin
````

val authorizer = IamAuthorizer { ApolloAmplifyConnector.signAppSyncRequest(it, "us-east-1") }
```

</InlineFilter>

### Connecting Amplify Data to Apollo client

Before you begin, you will need an Amplify Data backend deploy. To get started, see [Set up Data](/[platform]/swift/build-a-backend/data/set-up-data/). 

> Once you have deployed your backend and created the `amplify_outputs.json`, you can use Amplify library to read and retrieve your configuration values with the following steps:

Reviewer 3: ```suggestion
Once you have deployed your backend and created the `amplify_outputs.json` file, you can use Amplify library to read and retrieve your configuration values with the following steps:
```

                                                                store: store)
   // 3. Use `configuration.endpoint` with RequestChainNetworkTransport
    let transport = RequestChainNetworkTransport(interceptorProvider: interceptorProvider,
                                                endpointURL: configuration.endpoint)

    return ApolloClient(networkTransport: transport, store: store)
}
```

</InlineFilter>

Depending on your authorization strategy defined on your schema, you can use the corresponding Authorizer. To read more about the strategies and their corresponding auth modes, see [Available authorization strategies](/[platform]/build-a-backend/data/customize-authz/#available-authorization-strategies).

Some common ones are

*  `publicAPIkey` strategy, `apiKey` authMode, **APIKeyAuthorizer**
*  `guest` strategy, `identityPool` authMode, **IAMAuthorizer**
*  `owner` strategy, `userPool` authMode, **AuthTokenAuthorizer**

If you define multiple authorization strategies on a single model, you will have to create separate Apollo client instances for each Authorizer that you want to use in your app.

### Downloading the AWS AppSync schema

> The schema is used by Apollo’s code generation tool to generate API code that helps you execute GraphQL operations. To retrieve your AWS AppSync schema:

Reviewer 3: I wonder if we should include a programmatic example (despite not having a command to do this in the backend cli). In development these schema's will change quite often and downloading each time is cumbersome

Author: i think the base example should exist, as a way to show a simple way to get the schema from the console, while your suggestion to extend the example makes sense. The ideal example should utilize Apollo's codegen configuration to provide it the introspection endpoint and let Apollo retrieve it as part of the codegen process. 

Other programmatic examples would involve taking on a dependency on AWS SDK, like for Javascript and writing a script to download the schema using the appsync client

Author: I added an example for swift for fetching the introspection schema programmatically using `apollo-ios-cli`

<InlineFilter filters={["swift"]}>

> 1. Navigate to your API on the AWS AppSync console

Reviewer 3: ```suggestion
1. Navigate to your API on the [AWS AppSync console](https://console.aws.amazon.com/appsync/home)
```

2. On the left side, select Schema
3. When viewing your schema, there should a “Export schema” drop down. Select this and download the `schema.json` file.
4. Add this file to your project as directed by [Apollo Code Generation documentation](https://www.apollographql.com/docs/ios/code-generation/introduction) 
</InlineFilter>

<InlineFilter filters={["android"]}>

> 1. Navigate to your API on the AWS AppSync console

Reviewer 3: ```suggestion
1. Navigate to your API on the [AWS AppSync console](https://console.aws.amazon.com/appsync/home)
```

2. On the left side, select Schema
3. When viewing your schema, there should a “Export schema” drop down. Select this and download the `schema.json` file.
4. Add this file to your project as directed by [Apollo Code Generation documentation](https://www.apollographql.com/docs/ios/code-generation/introduction) 
</InlineFilter>

## Connecting to AWS AppSync real-time endpoint

The following example shows how you can create an Apollo client that allows performing GraphQL subscription operations with AWS AppSync. 

<InlineFilter filters={["swift"]}>

> ```

Reviewer 3: ````suggestion
```swift
````

  platforms: [
    'android',
    'swift',
  ]
};

export function getStaticPaths() {
  return getCustomStaticPath(meta.platforms);
}

export function getStaticProps(context) {
  return {
    props: {
      platform: context.params.platform,
      meta
    }
  };
}

AWS AppSync Apollo Extensions provides a seamless way to connect to your AWS AppSync while using the Apollo client. Apollo client is an open-source GraphQL client. 

<InlineFilter filters={["swift"]}>

To learn more about Apollo, see https://www.apollographql.com/docs/ios/.

</InlineFilter>

<InlineFilter filters={["android"]}>

To learn more about Apollo, see https://www.apollographql.com/docs/kotlin.

</InlineFilter>

### Features

AWS AppSync Apollo Extensions provide AWS AppSync authorizers to be used with the Apollo client to make it simple to configure runtime interceptors and apply the correct authorization payloads to your GraphQL operations.

The Amplify library provides components to facilitate configuring the authorizers with Apollo client by providing configuration values to connect to your Amplify Data backend.

<InlineFilter filters={["android"]}>

### Install the AWS AppSync Apollo library

> Add AWS AppSync Apollo Extensions dependency to your app/build.gradle.kts file:

Author: ```suggestion
Add `apollo-appsync` dependency to your app/build.gradle.kts file:
```

```kotlin title="app/build.gradle.kts"
dependencies {
    // highlight-start

>     implementation("com.amplifyframework:aws-appsync-apollo-extensions-android:1.0.0")    

Author: ```suggestion
    implementation("com.amplifyframework:apollo-appsync:1.0.0")    
```

```swift
import Apollo
import ApolloAPI
import AWSPluginsCore
import AWSAppSyncApolloExtensions

func createApolloClient() throws -> ApolloClient {
   let store = ApolloStore(cache: InMemoryNormalizedCache())
   
    // 1. Read AWS AppSync API configuration from `amplify_outputs.json`
    let configuration = try AWSAppSyncConfiguration(with: .amplifyOutputs)
    
   // 2. Use `configuration.apiKey` with APIKeyAuthorizer
    let authorizer = APIKeyAuthorizer(apiKey: configuration.apiKey ?? "")
    let interceptor = AppSyncInterceptor(authorizer)
    let interceptorProvider = DefaultPrependInterceptorProvider(interceptor: interceptor,
                                                                store: store)
   // 3. Use `configuration.endpoint` with RequestChainNetworkTransport
    let transport = RequestChainNetworkTransport(interceptorProvider: interceptorProvider,
                                                endpointURL: configuration.endpoint)

    return ApolloClient(networkTransport: transport, store: store)
}
```

</InlineFilter>

<InlineFilter filters={["android"]}>

> Add Amplify AppSync Extensions dependency to your app/build.gradle.kts file:

Author: ```suggestion
Add `apollo-appsync-amplify` dependency to your app/build.gradle.kts file:
```

```kotlin title="app/build.gradle.kts"
dependencies {
    // highlight-start

>     implementation("com.amplifyframework:amplify-appsync-apollo:1.0.0")    

Author: ```suggestion
    implementation("com.amplifyframework:apollo-appsync-amplify:1.0.0")    
```

    // highlight-end
}
```

```kotlin

> Code Sample

Author: ```suggestion
// Use CUP auth mode, reading configuration from AmplifyOutputs
val connector = ApolloAmplifyConnector(context, AmplifyOutputs(R.raw.amplify_outputs))
val apolloClient = ApolloClient.Builder()
    .appSync(connector.endpoint, connector.apiKeyAuthorizer())
    .build()
```

   
    // 1. Read AWS AppSync API configuration from `amplify_outputs.json`
    let configuration = try AWSAppSyncConfiguration(with: .amplifyOutputs)
    
   // 2. Use `configuration.apiKey` with APIKeyAuthorizer
    let authorizer = APIKeyAuthorizer(apiKey: configuration.apiKey ?? "")
    let interceptor = AppSyncInterceptor(authorizer)
    let interceptorProvider = DefaultPrependInterceptorProvider(interceptor: interceptor,
                                                                store: store)
   // 3. Use `configuration.endpoint` with RequestChainNetworkTransport
    let transport = RequestChainNetworkTransport(interceptorProvider: interceptorProvider,
                                                endpointURL: configuration.endpoint)

    return ApolloClient(networkTransport: transport, store: store)
}
```

</InlineFilter>

<InlineFilter filters={["android"]}>

Add `apollo-appsync-amplify` dependency to your app/build.gradle.kts file:

```kotlin title="app/build.gradle.kts"
dependencies {
    // highlight-start
    implementation("com.amplifyframework:apollo-appsync-amplify:1.0.0")    
    // highlight-end
}
```

```kotlin

> // Use CUP auth mode, reading configuration from AmplifyOutputs

Author: ```suggestion
// Use apiKey auth mode, reading configuration from AmplifyOutputs
```

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {
  title: 'AWS AppSync Apollo Extensions',
  description:
    'AWS AppSync Apollo Extensions',
  platforms: [
    'android',
    'swift',
  ]
};

export function getStaticPaths() {
  return getCustomStaticPath(meta.platforms);
}

export function getStaticProps(context) {
  return {
    props: {
      platform: context.params.platform,
      meta
    }
  };
}

> AWS AppSync Apollo Extensions provide a seamless way to connect to your AWS AppSync backend using Apollo client, an open-source GraphQL client. 

Reviewer 2: can you find and replace lowercase e for Extensions in this - this isn't an official name so needs to be lowercase (at least I haven't seen the naming doc :))

<InlineFilter filters={["swift"]}>

To learn more about Apollo, see https://www.apollographql.com/docs/ios/.

</InlineFilter>

<InlineFilter filters={["android"]}>

To learn more about Apollo, see https://www.apollographql.com/docs/kotlin.

</InlineFilter>

## Features

AWS AppSync Apollo Extensions provide AWS AppSync authorizers to be used with the Apollo client to make it simple to apply the correct authorization payloads to your GraphQL operations.

> The Amplify library provides components to facilitate configuring the authorizers with Apollo client by providing configuration values to connect to your Amplify Data backend.

Reviewer 2: ```suggestion
The Amplify library provides components to facilitate configuring the authorizers with Apollo client by providing configuration values to connect to your Amplify data backend.
```

#### API_KEY

<InlineFilter filters={["swift"]}>

```swift
import AWSAppSyncApolloExtensions

let authorizer = APIKeyAuthorizer(apiKey: "[API_KEY]")
let interceptor = AppSyncInterceptor(authorizer)
```

</InlineFilter>

<InlineFilter filters={["android"]}>

An `ApiKeyAuthorizer` can be used with a hardcoded API key, by fetching the key from some source, or reading it from `amplify_outputs.json`:

```kotlin
// highlight-start
// Use a hard-coded API key
val authorizer = ApiKeyAuthorizer("[API_KEY]")
//highlight-end
// or
// highlight-start
// Fetch the API key from some source. This function may be called many times,
// so it should implement appropriate caching internally.
val authorizer = ApiKeyAuthorizer { fetchApiKey() }
//highlight-end
// or
// highlight-start
// Using ApolloAmplifyConnector to read API key from amplify_outputs.json
val connector = ApolloAmplifyConnector(context, AmplifyOutputs(R.raw.amplify_outputs))
val authorizer = connector.apiKeyAuthorizer()
//highlight-end
```

</InlineFilter>

#### AMAZON_COGNITO_USER_POOLS

<InlineFilter filters={["swift"]}>

> If you are using Amplify Auth, you can create a method that retrieves the Cognito access token

Reviewer 2: Amplify auth

Reviewer: We use capitalized category names elsewhere in our documentation (for example, [here](https://docs.amplify.aws/android/build-a-backend/auth/set-up-auth/) and [here](https://docs.amplify.aws/android/build-a-backend/data/set-up-data/))

let interceptor = AppSyncInterceptor(authorizer)
```
</InlineFilter>

<InlineFilter filters={["android"]}>

You can use `AmplifyApolloConnector` to get an `AuthTokenAuthorizer` instance that supplies the token for the current logged-in Amplify user, or implement the token fetching yourself.

```kotlin
// highlight-start
// Using ApolloAmplifyConnector to get the authorizer that connects to your 
// Amplify instance
val connector = ApolloAmplifyConnector(context, AmplifyOutputs(R.raw.amplify_outputs))
val authorizer = connector.authTokenAuthorizer()
//highlight-end
// or
// highlight-start
// Using the ApolloAmplifyConnector companion function
val authorizer = AuthTokenAuthorizer { 
  ApolloAmplifyConnector.fetchLatestCognitoAuthToken() 
}
//highlight-end
// or
// highlight-start
// Use your own token fetching. This function may be called many times,
// so it should implement appropriate caching internally.
val authorizer = AuthTokenAuthorizer {
  fetchLatestAuthToken()
}
//highlight-end
```

</InlineFilter>

You can provide your own custom `fetchLatestAuthToken` provider for **AWS_LAMBDA** and **OPENID_CONNECT** auth modes. 

#### AWS_IAM

<InlineFilter filters={["swift"]}>

> If you are using Amplify Auth, you can use the following method for AWS_IAM auth

Reviewer 2: Amplify auth

```swift
import AWSCognitoAuthPlugin
import AWSAppSyncApolloExtensions

let authorizer = IAMAuthorizer(
    signRequest: AWSCognitoAuthPlugin.createAppSyncSigner(
        region: "[REGION]"))
```

</InlineFilter>

<InlineFilter filters={["android"]}>

If you are using `apollo-appsync-amplify`, you can use the `ApolloAmplifyConnector` to delegate token fetching and request 
signing to Amplify.

```kotlin
// highlight-start
// Using ApolloAmplifyConnector to get the authorizer that connects to your 
// Amplify instance
val connector = ApolloAmplifyConnector(context, AmplifyOutputs(R.raw.amplify_outputs))
val authorizer = connector.iamAuthorizer()
//highlight-end
// or
// highlight-start
// Using the ApolloAmplifyConnector companion function
val authorizer = IamAuthorizer { 
  ApolloAmplifyConnector.signAppSyncRequest(it, "us-east-1") 
}
//highlight-end
```

</InlineFilter>

### Connecting Amplify Data to Apollo client

> Before you begin, you will need an Amplify Data backend deploy. To get started, see [Set up Data](/[platform]/build-a-backend/data/set-up-data/). 

Reviewer 2: Amplify data

dependencies {
    // highlight-start
    implementation("com.amplifyframework:apollo-appsync-amplify:1.0.0")    
    // highlight-end
}
```

```kotlin
// Use apiKey auth mode, reading configuration from AmplifyOutputs
val connector = ApolloAmplifyConnector(context, AmplifyOutputs(R.raw.amplify_outputs))
val apolloClient = ApolloClient.Builder()
    .appSync(connector.endpoint, connector.apiKeyAuthorizer())
    .build()
```

</InlineFilter>

Depending on your authorization strategy defined on your schema, you can use the corresponding Authorizer. To read more about the strategies and their corresponding auth modes, see [Available authorization strategies](/[platform]/build-a-backend/data/customize-authz/#available-authorization-strategies).

Some common ones are

*  `publicAPIkey` strategy, `apiKey` authMode, **APIKeyAuthorizer**
*  `guest` strategy, `identityPool` authMode, **IAMAuthorizer**
*  `owner` strategy, `userPool` authMode, **AuthTokenAuthorizer**

If you define multiple authorization strategies on a single model, you will have to create separate Apollo client instances for each Authorizer that you want to use in your app.

### Downloading the AWS AppSync schema

> The schema is used by Apollo’s code generation tool to generate API code that helps you execute GraphQL operations. The following steps integrate your AppSync schema with Apollo's code generation process:

Reviewer 2: AppSync is prefix required - always need AWS before it (AWS AppSync)
